File: compiler/python/state-buffer-estimate/schedule/layerlevel.py
from utils.consts    import  *
import layers.layer
#from layers.layer import Layer

##########################################################
class LayerLevel(object):
    #-----------------------------------------------------------------
    def __init__(self, levelNum, initLayers):
        isinstance(initLayers, list)
        self.__LevelNum = levelNum
        self.__Layers = list(initLayers)

    #-----------------------------------------------------------------
    def remove(self, layer):
        assert(self.qContainsLayer(layer))
        self.__Layers.remove(layer)

    #-----------------------------------------------------------------
    def append(self, layer):
        assert(isinstance(layer, layers.layer.Layer))
        self.__Layers.append(layer)

    #-----------------------------------------------------------------
    def gLevelNum(self):
        return self.__LevelNum

    #-----------------------------------------------------------------
    def gLayers(self):
        return iter(self.__Layers)

    #-----------------------------------------------------------------
    def qContainsLayer(self, layer):
        return layer in self.__Layers

    #-----------------------------------------------------------------
    def gNumberLayers(self):
        return len(self.__Layers)

    #-----------------------------------------------------------------
    def qDataLevel(self):
        for layer in self.__Layers:
            if not layer.qDataLayer():
                return False
        return True

File: compiler/be/schedule/src/layerlevel.cpp
#include <algorithm>

#include "consts.hpp"
#include "layer.hpp"
#include "layerlevel.hpp"

namespace kcc {

namespace schedule {

LayerLevel::LayerLevel(kcc_int32 levelNum, const std::vector<layers::Layer*>& initLayers)
    : m_LevelNum(levelNum)
    , m_Layers(initLayers)
{}


//--------------------------------------------------------
void
LayerLevel::remove(layers::Layer* layer)
{
    //assert(qContainsLayer(layer));
    auto it = find(m_Layers.begin(), m_Layers.end(), layer);
    assert(it != m_Layers.end());
    auto lastIt = m_Layers.end() - 1;
    std::swap(*it, *lastIt);
    m_Layers.pop_back();
}

//--------------------------------------------------------
void
LayerLevel::append(layers::Layer* layer)
{
    m_Layers.push_back(layer);
}

//--------------------------------------------------------
bool
LayerLevel::qDataLevel() const
{
    for (auto layer : m_Layers) {
        if (!layer->qInputLayer()) {
            return false;
        }
    }
    return true;
}

//--------------------------------------------------------
bool
LayerLevel::qContainsLayer(layers::Layer* layer) const
{
    return std::find(m_Layers.begin(), m_Layers.end(), layer) != m_Layers.end();
}

}}

